Add intermediary data server for shuffle

jihoonson opened the pull request:
First PR for https://github.com/apache/incubator-druid/issues/8061.

### Description

This PR adds intermediary data server functionality to middleManagers. Documents for new configurations will be added once full functionalities for native parallel batch with shuffle is implemented.

This PR has:
- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added unit tests or modified existing tests to cover new code paths.
- [x] been tested in a test Druid cluster.

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/common/config/TaskConfig.java:
```diff
@@ -89,6 +94,13 @@ public TaskConfig(
                                 : directoryLockTimeout;
+    if (intermediarySegmentsLocations == null) {
+      this.intermediarySegmentsLocations = Collections.singletonList(
+          new StorageLocationConfig(new File(System.getProperty("java.io.tmpdir"), "intermediary-segments"), null, null)
```
```suggestion
          new StorageLocationConfig(new File(defaultDir(null, "intermediary-segments")), null, null)
```

jihoonson replied:
👍

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java:
```diff
@@ -0,0 +1,270 @@
+          }
+
+          if (exception != null) {
+            throw exception;
```
this code is running in a scheduler , why do we throw the overall exception in the end instead of logging them as they happened ?

himanshug replied:
can't we log the exceptions as they occur instead of accumulating all of them inside one and then logging it in the end ?

jihoonson replied:
Ah sure, fixed.

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java:
```diff
@@ -0,0 +1,270 @@
+    if (supervisorTaskChecker != null) {
+      supervisorTaskChecker.shutdownNow();
+      supervisorTaskChecker.awaitTermination(10, TimeUnit.SECONDS);
+      supervisorTaskChecker = null;
```
what is the use of setting it null here ?

himanshug replied:
it piqued my interest as usually you would do this sort of thing if you wanted to make something eligible for GC if it was a big object and part of large method so as to let it be GC'd asap or if you wanted to do something along the lines of making `IntermediaryDataManager` restartable etc.
I was trying to understand why you did that, but sounds like there is no reason so better to remove it :)

jihoonson replied:
Removed.

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java:
```diff
@@ -0,0 +1,270 @@
+   * Write a segment into one of configured locations. The location to write is chosen in a round-robin manner per
+   * supervisorTaskId.
+   *
+   * This method is only useful for the new Indexer model, and must not be called when tasks are running in the existing
```
does this mean parallel index task with shuffle support would not work with existing MMs?

why can't task processes have intermediary segment location propagated to them so that they could write data there and whole thing works for existing MMs as well ?

himanshug replied:
can you add what that other method is/would-be ? ... my understanding is that this method could be made static and peon process would write data by calling this method as well.

jihoonson replied:
Add a new static method which could be used by MMs.

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/http/ShuffleResource.java:
```diff
@@ -0,0 +1,116 @@
+  @Path("/task/{supervisorTaskId}")
+  public Response deletePartitions(
+      @PathParam("supervisorTaskId") String supervisorTaskId,
+      @QueryParam("dataSource") String dataSource
```
unused ?

himanshug replied:
ok... but why does it need to be declared here ? I understand that url needs that in query param for the filter to work but we probably don't need to capture that param in here.

jihoonson replied:
Hmm, I just removed dataSource and changed to use `StateResourceFilter` instead. Added the below javadoc.

> We use StateResourceFilter here because it performs an admin-like authorization and all endpoints here are supposed to be used for only internal communcation. Another possible alternate could be performing datasource-level authorization as in TaskResourceFilter. However, datasource information is not available in middleManagers or indexers yet which makes hard to use it. We could develop a new ResourceFileter in the future if needed.

clintropolis commented on server/src/main/java/org/apache/druid/server/http/security/TaskShuffleResourceFilter.java:
```diff
@@ -0,0 +1,72 @@
+ * This resource filter is used for data shuffle between native parallel index tasks. See ShuffleResource for details.
+ *
+ * It currently performs the authorization check for DATASOURCE, but ideally, it should be the authorization check
+ * for task data. This issue should be addressed in the future.
```
Could you clarify in these javadocs that I think you are referring to this should be doing what `TaskResourceFilter` is doing?

jihoonson replied:
Removed this class now in favor of using `StateResourceFilter`.

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java:
```diff
@@ -0,0 +1,274 @@
+                deletePartitions(supervisorTaskId);
+              }
+              catch (IOException e) {
+                if (exception == null) {
```
can't we log the exceptions as they occur instead of accumulating all of them inside one and then logging it in the end ?

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java:
```diff
@@ -0,0 +1,274 @@
+            }
+          }
+
+          final Map<String, TaskStatus> taskStatuses = indexingServiceClient.getTaskStatuses(expiredSupervisorTasks);
```
this has potential to throw exceptions and that would suppress further execution of this task... we should be stopping early only if the exception happened due to `interrupt`

himanshug replied:
it is usually a good idea to put scheduled tasks inside a big try catch to avoid premature exit due to unknown exceptions.

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java:
```diff
@@ -0,0 +1,274 @@
+            }
+          }
+
+          final Map<String, TaskStatus> taskStatuses = indexingServiceClient.getTaskStatuses(expiredSupervisorTasks);
```
this has potential to throw exceptions and that would suppress further execution of this task... we should be stopping early only if the exception happened due to interrupt

himanshug replied:
usually it would be good to put scheduled task inside a bit try-catch to avoid accidental premature exit or at least log what happened.

jihoonson replied:
Sounds good. Added.

jihoonson commented on indexing-service/src/test/java/org/apache/druid/indexing/worker/IntermediaryDataManagerManualAddAndDeleteTest.java:
```diff
@@ -0,0 +1,147 @@
+  }
+
+  @Test
+  public void testAddSegment() throws IOException
```
Hmm, removed.

jihoonson commented:
@himanshug and @clintropolis, thank you for taking a look. I addressed your comments. 

I also noticed that the state of `StorageLocation` was not updated properly before. I refactored `StorageLocation` to be a bit more decoupled with `DataSegment` and fixed it to update its state properly.

clintropolis commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java:
```diff
@@ -0,0 +1,310 @@
+    supervisorTaskChecker = Execs.scheduledSingleThreaded("intermediary-data-manager-%d");
+    // Discover partitions for new supervisorTasks
+    supervisorTaskChecker.scheduleAtFixedRate(
+        () -> {
```
This maybe seems large and complex enough to split into it's own method

jihoonson replied:
Extracted as a method.

clintropolis commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java:
```diff
@@ -0,0 +1,310 @@
+    // Note that the overlord sends a cleanup request when a supervisorTask is finished. The below check is to trigger
+    // the self-cleanup for when middleManager misses the cleanup request from the overlord.
+    supervisorTaskChecker.scheduleAtFixedRate(
+        () -> {
```
Same comment about splitting lambda into a method

jihoonson replied:
Extracted as a method.

clintropolis commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java:
```diff
@@ -0,0 +1,310 @@
+    throw new ISE("Can't find location to handle segment[%s]", segment);
+  }
+
+  public List<File> findPartitionFiles(String supervisorTaskId, Interval interval, int partitionId)
```
nit: could you put this and the non-static method that follows up with the other non-static methods?

jihoonson replied:
Moved around to put non-static methods together.

clintropolis commented:
[CI failures look legitimate](https://travis-ci.org/apache/incubator-druid/jobs/560213743#L969):
```
[ERROR] Failures: 
[ERROR] org.apache.druid.segment.loading.StorageLocationTest.testStorageLocation(org.apache.druid.segment.loading.StorageLocationTest)
[ERROR]   Run 1: StorageLocationTest.testStorageLocation:79->verifyLoc:99 expected:<990> but was:<980>
[ERROR]   Run 2: StorageLocationTest.testStorageLocation:79->verifyLoc:99 expected:<990> but was:<980>
[ERROR]   Run 3: StorageLocationTest.testStorageLocation:79->verifyLoc:99 expected:<990> but was:<980>
[ERROR]   Run 4: StorageLocationTest.testStorageLocation:79->verifyLoc:99 expected:<990> but was:<980>
```

jihoonson commented:
@clintropolis thanks, fixed the test failure.

himanshug commented on indexing-service/src/main/java/org/apache/druid/indexing/worker/IntermediaryDataManager.java:
```diff
@@ -0,0 +1,328 @@
+          }
+          catch (Exception e) {
+            log.warn(e, "Error while cleaning up partitions for expired supervisors");
+          }
```
need to exit on  "InterruptedException" or e.cause being InterruptedException  

warn mean no action needed, but in this case user might need to come in and do manual cleanup , so error level log is preferable.

himanshug replied:
actually it might be better for `deleteExpiredSuprevisorTaskPartitionsIfNotRunning()` to throw InterruptedException so that code here doesn't need to concern itself with e.cause check as that is internal detail of that method.

jihoonson replied:
Oh yeah, forgot to handle that. Thanks. 👍

jihoonson replied:
Just changed `indexingServiceClient.getTaskStatuses` to throw `InterruptedException`. I think it's better since it's a blocking call.

himanshug replied:
sorry, now that I see the change .. I realized it wasn't strictly necessary in this context ... but nice to have.

jihoonson replied:
Yeah, I think it's better to log as an error rather than warning.

jihoonson commented:
@himanshug thanks for the review!
